google_guest_agent/agentcrypto/mtls_mds.go

```diff
@@ -131,62 +120,71 @@ func (j *CredsJob) extractKey(importBlob *tpm.ImportBlob) ([]byte, error) {
 	return dek, nil
 }
 
-// fetchAndWriteClientCredentials fetches encrypted client credentials from MDS,
+// fetchClientCredentials fetches encrypted client credentials from MDS,
 // extracts Key Encryption Key (KEK) from vTPM, decrypts the client credentials using KEK,
-// verifies these credentials are signed by root CA and writes it to the output file.
-func (j *CredsJob) fetchAndWriteClientCredentials(ctx context.Context, rootCA, outputFile string) error {
+// and verifies that the certificate is signed by root CA.
+func (j *CredsJob) fetchClientCredentials(ctx context.Context, rootCA string) ([]byte, error) {
 	resp, err := j.getClientCredentials(ctx)
 	if err != nil {
-		return err
+		return []byte{}, err
 	}
 
 	dek, err := j.extractKey(resp.GetKeyImportBlob())
 	if err != nil {
-		return err
+		return []byte{}, err
 	}
 
 	plaintext, err := decrypt(dek, resp.GetEncryptedCredentials(), nil)
 	if err != nil {
-		return err
+		return []byte{}, err
 	}
 
 	if err := verifySign(plaintext, rootCA); err != nil {
-		return err
+		return []byte{}, err
 	}
 
-	if err := os.WriteFile(outputFile, plaintext, 0644); err != nil {
-		return fmt.Errorf("unable to write client credentials to file: %w", err)
-	}
-
-	logger.Infof("Successfully wrote client credentials to %q", outputFile)
-	return nil
+	return plaintext, nil
 }
 
 // Run generates the required credentials for MTLS MDS workflow.
 //
 // 1. Fetches, verifies and writes Root CA cert from UEFI variable to /etc/pki/tls/certs/mds/root.crt
 // 2. Fetches encrypted client credentials from MDS, decrypts it via vTPM and writes it to /etc/pki/tls/certs/mds/client.key
 //
+// Note that these credentials are at `C:\Program Files\Google\Compute Engine\certs\mds` on Windows.
+// Additionally agent also generates a PFX file on windows that can be used invoking HTTPS endpoint.
+//
 // Example usage of these credentials to call HTTPS endpoint of MDS:
 //
 // curl --cacert /etc/pki/tls/certs/mds/root.crt -E /etc/pki/tls/certs/mds/client.key -H "MetadataFlavor: Google" https://169.254.169.254
+//
+// Windows example:
+//
+// $cert = Get-PfxCertificate -FilePath "C:\Program Files\Google\Compute Engine\certs\mds\client.key.pfx"
+// or
+// $cert = Get-ChildItem Cert:\LocalMachine\My | Where-Object { $_.Issuer -like "*google.internal*" }
+// Invoke-RestMethod -Uri https://169.254.169.254 -Method Get -Headers @{"Metadata-Flavor"="Google"} -Certificate $cert
 func (j *CredsJob) Run(ctx context.Context) (bool, error) {
-	// defaultCredsDir is the directory location for MTLS MDS credentials.
-	defaultCredsDir := "/etc/pki/tls/certs/mds"
+	logger.Infof("Fetching Root CA cert...")
 
-	// TODO: Finalize on where to store certificates on windows.
-	if runtime.GOOS == "windows" {
-		defaultCredsDir = `C:\Users`
+	v, err := j.readRootCACert(googleRootCACertUEFIVar)
```

Comment:
Can you add a EUFI variable check in the begging of this function and if the variable is not set we just return? That way we return early and use that as a signal that the feature is not enable?

Reply:
You can do it in a follow-up PR though.